src/rdb.c

```diff
@@ -1918,22 +1918,26 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, int dbid) {
                     zsetConvert(o,OBJ_ENCODING_SKIPLIST);
                 break;
             case RDB_TYPE_HASH_ZIPLIST:
-                if (deep_integrity_validation) server.stat_dump_payload_sanitizations++;
-                if (!hashZiplistValidateIntegrity(encoded, encoded_len, deep_integrity_validation)) {
-                    rdbReportCorruptRDB("Hash ziplist integrity check failed.");
-                    zfree(encoded);
-                    o->ptr = NULL;
-                    decrRefCount(o);
-                    return NULL;
-                }
-                o->type = OBJ_HASH;
-                o->encoding = OBJ_ENCODING_ZIPLIST;
-                if (ziplistLen(o->ptr) > server.hash_max_listpack_entries) {
-                    hashTypeConvert(o, OBJ_ENCODING_HT);
-                } else {
-                    hashTypeConvert(o, OBJ_ENCODING_LISTPACK);
+                {
+                    unsigned char *lp = lpNew(encoded_len);
+                    if (!hashZiplistValidateIntegrity(encoded, encoded_len, &lp)) {
```

Comment:
If the main purpose of this call now is the conversion, let's rename it.
maybe `hashZiplistConvertAndValidate` or `hashZiplistConvertToListpack` and place a comment that:
1. it's safe to call on non-validated ziplists.
2. it returns false when encounter an integrity validation issue.

also, we now lost the ability to convert directly from ziplist to dict, maybe that's not that bad, it shouldn't normally happen.

Reply:
Fixed